### PR TITLE
[Banner]  Measure `contentNode` on change rather than `window.onResize`

### LINKED ISSUE #11770: `Banner` icon loads at `center` and moves to `start` `align-items`
### Summary

When loading a page with a `Banner`, the icon on the left side starts with an `align-items: center` through the `--pc-inline-stack-block-align` property and after the page finishes loading it changes to `align-items: start`. It's not very visible in isolation, but it can be reproduced.

### Expected behavior

The alignment should be consistent to it doesn't move around, ideally staying with an `align-items: start`.

### Actual behavior

Wrap animated gifs/videos in a details tag:
<details>
  <summary>Problem occurring in our 1p app</summary>
 

https://github.com/Shopify/polaris/assets/15715/da40f32a-21e8-48eb-8f03-2d3b82861fd0

</details>

<details>
  <summary>Problem occurring in the sandbox</summary>

https://github.com/Shopify/polaris/assets/15715/6c45d260-1b94-47ac-ae38-6ed3c7df5673

(The above video shows the problem in CodeSandbox, but you'll need to download it and look at it frame by frame to see the issue.)
</details>

### Steps to reproduce

[Link to sandbox](https://codesandbox.io/p/sandbox/sharp-swirles-wgvntx?file=%2FApp.tsx%3A19%2C1-20%2C1)
1. Create `Banner` with enough text to take several lines
2. Reload the page with an empty cache
3. See the icon move from center to top

### Are you using React components?

Yes

### Polaris version number

_No response_

### Browser

_No response_

### Device

_No response_

### PR DESCRIPTION
**Pull request by LA1CH3**
### WHY are these changes introduced?

Closes https://github.com/Shopify/polaris/issues/11770 

This PR fixes an issue where the `blockAlign` value was not being properly updated when `contentNode.offsetHeight` would change. This was because the measurement/calculation would only trigger on `window.resize`, and would not trigger via other actions (such as a banner having content dynamically added/removed). 

### WHAT is this pull request doing?

I've fixed this by using `ResizeObserver` to trigger recalculation whenever `contentNode` is resized, rather than the current `handleResize` which is only triggered on `window.resize`.

### How to 🎩

[Storybook link](https://5d559397bae39100201eedc1-wzruvmwjwa.chromatic.com/?path=/story/all-components-banner--with-dynamic-content)
[Web w/ snapshot](https://admin.web.pf-pdp-required-sku-follows.jay-laiche.us.spin.dev/store/shop1/products/2)
 
This includes an example story where the banner content is dynamically changed, resulting in `contentNode` changes, and triggering the correct `blockAlign` property applied. 

### 🎩 checklist

- [ ] Tested a [snapshot](https://github.com/Shopify/polaris/blob/main/documentation/Releasing.md#-snapshot-releases)
- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Review by sophschneider — APPROVED** [review]
Thanks for fixing this! 🎉

Would you be able to `/snapit` and do a quick tophat in the admin? Storybook looked good.

Also I noticed there was a flash when expanding the content where there is a render of the new content before the icon is adjusted. Its not blocking but I wonder how we could fix this 🤔
https://github.com/Shopify/polaris/assets/20652326/4d4731ef-63ff-434a-b9f7-6583b8166d23

**Comment by LA1CH3** [comment]
@sophschneider I see that issue too 🤔 I have switched to use an unwrapped (as in, not wrapped in `useCallback`) callback ref for `contentNode` instead of a `ResizeObserver` and I no longer see the delay. I guess the only downside of this pattern is that the callback ref will run on every re-render, regardless if `contentNode` size has changed or not. Its probably a relatively inexpensive computation though, so maybe this is acceptable as it avoids layout shift. Thoughts?

You can view the [updated storybook link here](https://5d559397bae39100201eedc1-axdzrmdnhj.chromatic.com/?path=/story/all-components-banner--with-dynamic-content).

**Comment by LA1CH3** [comment]
@sophschneider I've included a web + snapshot link in the PR for tophatting in the admin. Lmk what you think about the callback ref approach.

File: .changeset/weak-crabs-arrive.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': patch
+---
+
+Use callback ref to more accurately detect content changes in Banner
```
Comment (chloerice) [inline .changeset/weak-crabs-arrive.md]:
```suggestion
Fixed `Banner` not accurately detecting content changes
```

**Review by chloerice — APPROVED** [review]
Works great @LA1CH3!

**Comment by LA1CH3** [comment]
Hey @chloerice so it actually looks like the issue was because when the callback ref was running on initial render, `iconNode` and `dismissIconNode` were both undefined- so the `blockAlign` was just defaulting to its initial value of `center`. 

I think the best way to tackle this in response to any of the elements changing is to use `ResizeObserver`. I tried this originally and there was/is a very brief delay in the `blockAlign` property changing; I think this is because the logic runs only after one of the nodes has already resized. I tried testing it again now and sometimes there is a delay, sometimes not. Do you mind testing again and lmk what you think?

**Comment by LA1CH3** [comment]
Closing in favor of https://github.com/Shopify/polaris-internal/pull/65

**Pull request closed without merging**